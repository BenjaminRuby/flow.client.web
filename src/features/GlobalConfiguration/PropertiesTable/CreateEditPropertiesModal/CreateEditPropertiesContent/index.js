import React, { Component } from "react";
import PropTypes from "prop-types";
import axios from "axios";
import { Formik } from "formik";
import * as Yup from "yup";
import { TextInput } from "carbon-components-react";
import Body from "@boomerang/boomerang-components/lib/ModalContentBody";
import ConfirmButton from "@boomerang/boomerang-components/lib/ModalConfirmButton";
import Footer from "@boomerang/boomerang-components/lib/ModalContentFooter";
import LoadingAnimation from "@boomerang/boomerang-components/lib/LoadingAnimation";
import { notify, Notification } from "@boomerang/boomerang-components/lib/Notifications";
import Toggle from "@boomerang/boomerang-components/lib/Toggle";
import INPUT_TYPES from "Constants/inputTypes";
import { BASE_SERVICE_URL } from "Config/servicesConfig";
import styles from "./createEditPropertiesContent.module.scss";

class CreateEditPropertiesContent extends Component {
  static propTypes = {
    addPropertyInStore: PropTypes.func,
    closeModal: PropTypes.func,
    handleEditClose: PropTypes.func,
    isEdit: PropTypes.bool,
    property: PropTypes.object,
    propertyKeys: PropTypes.array.isRequired,
    updatePropertyInStore: PropTypes.func
  };

  handleSubmit = (values, options) => {
    const { addPropertyInStore, isEdit, property, updatePropertyInStore } = this.props;
    const type = values.secured ? INPUT_TYPES.PASSWORD : INPUT_TYPES.TEXT;
    const newProperty = isEdit ? { ...values, type, id: property.id } : { ...values, type };
    const storeUpdate = isEdit ? updatePropertyInStore : addPropertyInStore;
    delete newProperty.secured;

    axios({
      method: isEdit ? "put" : "post",
      url: isEdit ? `${BASE_SERVICE_URL}/config/${newProperty.id}` : `${BASE_SERVICE_URL}/config`,
      data: newProperty
    })
      .then(response => {
        storeUpdate(newProperty);
        notify(
          <Notification
            type="success"
            title={isEdit ? "Property Updated" : "Property Created"}
            message={
              isEdit ? `Request to update ${newProperty.label} succeeded` : "Request to create property succeeded"
            }
          />
        );
        options.setSubmitting(false);
      })
      .then(() => {
        isEdit ? this.props.handleEditClose() : this.props.closeModal();
      })
      .catch(error => {
        notify(
          <Notification
            type="error"
            title={isEdit ? "Update Property Failed" : "Create Property Failed"}
            message={"Something went wrong"}
          />
        );
        options.setSubmitting(false);
      });
  };

  render() {
    const { isEdit, property, propertyKeys } = this.props;

    return (
      <Formik
        initialValues={{
          label: property ? property.label : "",
          description: property ? property.description : "",
          key: property ? property.key : "",
          value: property ? property.value : "",
          secured: property ? property.type === INPUT_TYPES.PASSWORD : false
        }}
        onSubmit={this.handleSubmit}
        validationSchema={Yup.object().shape({
          label: Yup.string().required("Enter a label"),
          key: Yup.string()
            .required("Enter a key")
            .notOneOf(propertyKeys, "Key must be unique"),
          value: Yup.string().required("Enter a value"),
          description: Yup.string(),
          secured: Yup.boolean()
        })}
      >
        {props => {
          const {
            values,
            touched,
            errors,
            isSubmitting,
            isValid,
            handleChange,
            handleBlur,
            handleSubmit,
            setFieldValue
          } = props;

          if (isSubmitting) {
            return <LoadingAnimation theme="bmrg-white" message="We'll be right with you" />;
          }

          return (
            <form onSubmit={handleSubmit}>
              <Body
                style={{
                  display: "block",
                  margin: "auto",
                  width: "60%",
                  height: "28rem",
                  overflowY: "auto"
                }}
              >
                <div className={styles.input}>
                  <TextInput
                    id="label"
                    labelText="Label"
                    placeholder="Label"
                    value={values.label}
                    onBlur={handleBlur}
                    onChange={handleChange}
                    invalid={errors.label && touched.label}
                    invalidText={errors.label}
                  />
                </div>
                <div className={styles.input}>
                  <TextInput
                    id="key"
                    labelText="key"
                    placeholder="Key"
                    value={values.key}
                    onBlur={handleBlur}
                    onChange={handleChange}
                    invalid={errors.key && touched.key}
                    invalidText={errors.key}
                  />
                </div>
                <div className={styles.input}>
                  <TextInput
                    id="description"
                    labelText="Description"
                    placeholder="Description"
                    value={values.description}
                    onBlur={handleBlur}
                    onChange={handleChange}
                  />
                </div>
                <div className={styles.input}>
                  <TextInput
                    id="value"
                    labelText="Value"
                    placeholder="Value"
                    value={values.value}
                    onBlur={handleBlur}
                    onChange={handleChange}
                    invalid={errors.value && touched.value}
                    invalidText={errors.value}
                  />
                </div>
                <div className={styles.toggleContainer}>
                  <label htmlFor="secured" className={styles.toggleLabel}>
                    Secured
                  </label>
                  <Toggle
                    className={styles.toggle}
                    id="secured"
                    checked={values.secured}
                    onChange={checked => setFieldValue("secured", checked)}
                    theme="bmrg-white"
                  />
                </div>
              </Body>
              <Footer style={{ flexDirection: "column", alignItems: "center", justifyContent: "flex-start" }}>
                <ConfirmButton
                  text={isEdit ? "SAVE" : "CREATE"}
                  type="submit"
                  disabled={!isValid || isSubmitting}
                  theme="bmrg-white"
                />
              </Footer>
            </form>
          );
        }}
      </Formik>
    );
  }
}

export default CreateEditPropertiesContent;
