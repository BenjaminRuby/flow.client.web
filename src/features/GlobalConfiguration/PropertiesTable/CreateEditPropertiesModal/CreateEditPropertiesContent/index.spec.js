import React from "react";
import CreateEditPropertiesContent from ".";
import renderer from "react-test-renderer";

const mockfn = jest.fn();
const props = {
  isEdit: true,
  property: { id: "1", key: "test.key", value: "testing" },
  propertyKeys: ["test"],
  handleEditClose: mockfn,
  addPropertyInStore: mockfn,
  updatePropertyInStore: mockfn
};

describe("CreateEditPropertiesContent --- Snapshot Test", () => {
  it("Capturing Snapshot of CreateEditPropertiesContent", () => {
    const component = renderer.create(<CreateEditPropertiesContent {...props} />).toJSON();
    expect(component).toMatchSnapshot();
  });
});
