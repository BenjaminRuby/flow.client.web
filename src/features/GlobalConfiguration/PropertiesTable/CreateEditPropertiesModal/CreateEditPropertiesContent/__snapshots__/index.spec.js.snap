// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateEditPropertiesContent --- Snapshot Test Capturing Snapshot of CreateEditPropertiesContent 1`] = `
<form
  onSubmit={[Function]}
>
  <main
    className="bmrg--c-modal-content-body"
    style={
      Object {
        "display": "block",
        "height": "28rem",
        "margin": "auto",
        "overflowY": "auto",
        "width": "60%",
      }
    }
  >
    <div
      className="input"
    >
      <div
        className="bx--form-item bx--text-input-wrapper"
      >
        <label
          className="bx--label"
          htmlFor="label"
        >
          Label
        </label>
        <div
          className="bx--text-input__field-wrapper"
          data-invalid={null}
        >
          <input
            className="bx--text-input bx--text__input"
            disabled={false}
            id="label"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            placeholder="Label"
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      className="input"
    >
      <div
        className="bx--form-item bx--text-input-wrapper"
      >
        <label
          className="bx--label"
          htmlFor="key"
        >
          key
        </label>
        <div
          className="bx--text-input__field-wrapper"
          data-invalid={null}
        >
          <input
            className="bx--text-input bx--text__input"
            disabled={false}
            id="key"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            placeholder="Key"
            type="text"
            value="test.key"
          />
        </div>
      </div>
    </div>
    <div
      className="input"
    >
      <div
        className="bx--form-item bx--text-input-wrapper"
      >
        <label
          className="bx--label"
          htmlFor="description"
        >
          Description
        </label>
        <div
          className="bx--text-input__field-wrapper"
          data-invalid={null}
        >
          <input
            className="bx--text-input bx--text__input"
            disabled={false}
            id="description"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            placeholder="Description"
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      className="input"
    >
      <div
        className="bx--form-item bx--text-input-wrapper"
      >
        <label
          className="bx--label"
          htmlFor="value"
        >
          Value
        </label>
        <div
          className="bx--text-input__field-wrapper"
          data-invalid={null}
        >
          <input
            className="bx--text-input bx--text__input"
            disabled={false}
            id="value"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            placeholder="Value"
            type="text"
            value="testing"
          />
        </div>
      </div>
    </div>
    <div
      className="toggleContainer"
    >
      <label
        className="toggleLabel"
        htmlFor="secured"
      >
        Secured
      </label>
      <div
        className="toggle --bmrg-flow"
        style={
          Object {
            "MozTransition": "opacity 0.25s",
            "MozUserSelect": "none",
            "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
            "WebkitTransition": "opacity 0.25s",
            "WebkitUserSelect": "none",
            "borderRadius": 10,
            "display": "inline-block",
            "msUserSelect": "none",
            "opacity": 1,
            "position": "relative",
            "textAlign": "left",
            "touchAction": "none",
            "transition": "opacity 0.25s",
            "userSelect": "none",
          }
        }
      >
        <div
          className="react-switch-bg"
          onClick={[Function]}
          onMouseDown={[Function]}
          style={
            Object {
              "MozTransition": "background 0.25s",
              "WebkitTransition": "background 0.25s",
              "background": "#486c89",
              "borderRadius": 10,
              "cursor": "pointer",
              "height": 20,
              "margin": 0,
              "position": "relative",
              "transition": "background 0.25s",
              "width": 40,
            }
          }
        />
        <div
          className="react-switch-handle"
          onClick={[Function]}
          onMouseDown={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "MozTransition": "background-color 0.25s, transform 0.25s, box-shadow 0.15s",
              "WebkitTransition": "background-color 0.25s, transform 0.25s, box-shadow 0.15s",
              "background": "#ff806c",
              "border": 0,
              "borderRadius": "50%",
              "boxShadow": null,
              "cursor": "pointer",
              "display": "inline-block",
              "height": 18,
              "outline": 0,
              "position": "absolute",
              "top": 1,
              "transform": "translateX(1px)",
              "transition": "background-color 0.25s, transform 0.25s, box-shadow 0.15s",
              "width": 18,
            }
          }
        />
        <input
          checked={false}
          disabled={false}
          id="secured"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyUp={[Function]}
          role="switch"
          style={
            Object {
              "border": 0,
              "clip": "rect(0 0 0 0)",
              "height": 1,
              "margin": -1,
              "overflow": "hidden",
              "padding": 0,
              "position": "absolute",
              "width": 1,
            }
          }
          type="checkbox"
        />
      </div>
    </div>
  </main>
  <footer
    className="bmrg--c-modal-content-footer"
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "column",
        "justifyContent": "flex-start",
      }
    }
  >
    <button
      className="bmrg--b-modal-confirm-button --disabled --bmrg-flow --disabled"
      disabled={true}
      type="submit"
    >
      SAVE
    </button>
  </footer>
</form>
`;
